### PR TITLE
Add California ID to client profile frontend (DEV-894)

### PR DESCRIPTION
**Pull request by mikefeldberg**
DEV-894

on hold until DEV-896 goes out

## Summary by Sourcery

Add support for California ID in client profiles, including input validation and duplicate checking.

New Features:
- Add support for California ID in client profiles, allowing users to input and validate California ID numbers.

Enhancements:
- Improve duplicate checking by integrating a query to check for existing client profiles with the same California ID.

## Summary by Sourcery

Add support for California ID in client profiles, including input validation and duplicate checking. Enhance the system to prevent duplicate entries by checking existing profiles for matching California IDs. Update tests to cover new functionality.

New Features:
- Add support for California ID in client profiles, allowing users to input and validate California ID numbers.

Enhancements:
- Improve duplicate checking by integrating a query to check for existing client profiles with the same California ID.

Tests:
- Add tests for searching client profiles by California ID to ensure correct duplicate detection.

### CODE REVIEW

File: libs/expo/betterangels/src/lib/screens/AddEditClient/PersonalInfo/index.tsx [hunk 1]
```diff
@@ -17,6 +21,49 @@ interface IPersonalInfoProps {
...
+    const shouldCheck =
+      californiaId &&
+      californiaId.length === californiaIdLength &&
+      !errors['californiaId'];
+
+    setCheckForDuplicates(shouldCheck);
+  }, [californiaId, errors]);
+
+  useEffect(() => {
+    if (data && data?.clientProfiles?.length > 0) {
```
Comment (mikefeldberg) [inline libs/expo/betterangels/src/lib/screens/AddEditClient/PersonalInfo/index.tsx:58]: alternatively, we could check if the current client == the returned client. Then we could remove the excluded id param. I think it's faster to do this work on the backend, so I went with that approach

File: apps/betterangels-backend/clients/types.py [hunk 2]
```diff
@@ -58,9 +58,11 @@ class ClientProfileOrder:
 
 @strawberry.input
 class ClientSearchInput:
-    first_name: str | None = None
-    last_name: str | None = None
-    middle_name: str | None = None
+    excluded_client_profile_id: Optional[str] = None
```
Comment (mikefeldberg) [inline apps/betterangels-backend/clients/types.py:61]: see: https://github.com/BetterAngelsLA/monorepo/pull/705/files#r1839631579

File: libs/expo/betterangels/src/lib/screens/AddEditClient/PersonalInfo/index.tsx [hunk 3]
```diff
@@ -17,6 +21,49 @@ interface IPersonalInfoProps {
...
+          excludedClientProfileId: (clientProfileId as string) || null,
+          californiaId: californiaId,
+        },
+      },
+    },
+    fetchPolicy: 'network-only',
+    nextFetchPolicy: 'network-only',
+  });
+
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/PersonalInfo/index.tsx:48]: seems unnecessary since we're only hitting the API when the contents are an exact length

File: libs/expo/betterangels/src/lib/hooks/californiaIdValidation/useCaliforniaIdValidation.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { useClientProfilesQuery } from '../../screens/Clients/__generated__/Clients.generated';
```
Comment (mikefeldberg) [inline libs/expo/betterangels/src/lib/hooks/californiaIdValidation/useCaliforniaIdValidation.ts:1]: maybe this needs its own query. something that _only_ returns the id

File: libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx [hunk 5]
```diff
@@ -206,6 +206,15 @@ export default function AddEditClient({ id }: { id?: string }) {
             message: 'User with this Email already exists.',
           });
           return;
+        } else if (
+          operationResult.messages[0].message ===
```
[the remark this answers is not in the document]
Comment (tglaz) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx:210]: and given the `operationResult.messages` returned payload (see below), we can maybe compare on `field` and `kind` only? I say this assuming that `VALIDATION` means a validation error.

```javascript
[
    {
        "__typename": "OperationMessage",
        "field": "californiaId",
        "kind": "VALIDATION",
        "message": "California ID must be 1 letter followed by 7 numbers"
    }
]
```

File: libs/expo/betterangels/src/lib/hooks/californiaIdValidation/useCaliforniaIdValidation.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+export default function useCaliforniaIdValidation(
```
Comment (tglaz) [inline libs/expo/betterangels/src/lib/hooks/californiaIdValidation/useCaliforniaIdValidation.ts:3]: I'd consider renaming as this just checks uniqueness part of validation and not full validation.
I'd also expect validation to return a boolean or a message etc.

File: libs/expo/betterangels/src/lib/screens/AddEditClient/PersonalInfo/index.tsx [hunk 7]
```diff
@@ -17,6 +21,25 @@ interface IPersonalInfoProps {
 
 export default function PersonalInfo(props: IPersonalInfoProps) {
   const { scrollRef, expanded, setExpanded } = props;
+  const { id: clientProfileId } = useLocalSearchParams();
+  const { watch, setError, clearErrors } = useFormContext();
+  const californiaId = watch('californiaId');
```
Comment (tglaz) [inline libs/expo/betterangels/src/lib/screens/AddEditClient/PersonalInfo/index.tsx:26]: I believe `watch` may trigger re-renders in parent and sibling components with each change, so kind of expensive. 
`useWatch` should be safer in this particular case

```javascript
  const { control, setError, clearErrors } = useFormContext();

  const [californiaId] = useWatch({
    control,
    name: ['californiaId'],
  });
```

**Review by tglaz — CHANGES_REQUESTED** [review]

File: libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx [hunk 8]
```diff
@@ -206,6 +206,15 @@ export default function AddEditClient({ id }: { id?: string }) {
             message: 'User with this Email already exists.',
           });
           return;
+        } else if (
```
[the remark this answers is not in the document]
Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx:209]: I think the error here is the `return`. there should be one `return` at the bottom of the block. updated

File: libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx:210]: I added this to the list of things we're looking for. Tho form validation should prevent this from being submitted. side note: maybe we should have an `errorMessagesMap` where all of these live

File: libs/expo/shared/static/src/lib/regex.ts [hunk 9]
```diff
@@ -4,4 +4,5 @@ export const Regex = {
   time: /^(?:[01]\d|2[0-3]):[0-5]\d$/,
   email: /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/,
   phoneNumber: /^[2-9]\d{2}[2-9]\d{6}$/,
+  californiaId: /^[A-Z]\d{7}$/,
```
[the remark this answers is not in the document]
Comment (mikefeldberg) [reply libs/expo/shared/static/src/lib/regex.ts:7]: the backend validation requires capitalized first letter. I added `autoCapitalize` to the input for good measure

File: libs/expo/betterangels/src/lib/screens/AddEditClient/PersonalInfo/index.tsx (hunk 7 above)
Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/PersonalInfo/index.tsx]: neat, thanks!

File: libs/expo/betterangels/src/lib/hooks/californiaIdValidation/useCaliforniaIdValidation.ts (hunk 4 above)
Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/hooks/californiaIdValidation/useCaliforniaIdValidation.ts]: @tglaz I added a slim query just for this. not sure if it should be colocated with the hook 🤷‍♂️

File: libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (tglaz) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx:205]: So perhaps i'm reading it wrong, but below is what I see the function doing.
Only 1 out of these blocks will ever execute, with thrown error being the fallback. 
The `if/else` chain obscures that in my view

```javascript
      if (
        operationResult?.__typename === 'OperationInfo' &&
        operationResult.messages.length > 0
      ) {
        const resultMessage = operationResult.messages[0].message;

        if (resultMessage === 'X') {
          methods.setError('user.email' ...

          return;
        } 
        
        if (resultMessage === 'Y') {
          methods.setError('californiaId' ...

          return;
        } 
        
        if (resultMessage === 'Z') {
          methods.setError('californiaId' ...

          return;
        } 
        
        throw new Error(
          `Failed to update a client profile: ${resultMessage}`
        );
      }
```

File: libs/expo/betterangels/src/lib/hooks/californiaIdUniqueCheck/useCaliforniaIdUniqueCheck.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import { Regex } from '@monorepo/expo/shared/static';
+import { useClientCaliforniaUniqueCheckQuery } from './__generated__/ClientCaliforniaUniqueCheck.generated';
+
+export default function useCaliforniaIdUniqueCheck(
+  californiaId: string,
+  clientProfileId?: string
+) {
+  const { data } = useClientCaliforniaUniqueCheckQuery({
+    skip: !californiaId || !Regex.californiaId.test(californiaId),
```
Comment (tglaz) [inline libs/expo/betterangels/src/lib/hooks/californiaIdUniqueCheck/useCaliforniaIdUniqueCheck.ts:9]: just the regex would be enough: `skip: !Regex.californiaId.test(californiaId)`

File: libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (tglaz) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx]: form validation is completely separate, and better to turn it off (ignore) when testing this functionality

As for `errorMessagesMap` yeah, think you're right and we should communicate those in a better way.
I wonder if we could specify the type of validation failed and return that with the payload.

File: libs/expo/shared/static/src/lib/regex.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (tglaz) [reply libs/expo/shared/static/src/lib/regex.ts:7]: ok, though this may cause some confusion down the road

File: libs/expo/betterangels/src/lib/hooks/californiaIdValidation/useCaliforniaIdValidation.ts (hunk 4 above)
Comment (tglaz) [reply libs/expo/betterangels/src/lib/hooks/californiaIdValidation/useCaliforniaIdValidation.ts]: I'd go with whatever we're doing today I suppose
I'm not used to doing that, but I realize there are some graphQl conventions etc that would say yes.

File: libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx:205]: wfm

File: libs/expo/betterangels/src/lib/hooks/californiaIdUniqueCheck/useCaliforniaIdUniqueCheck.ts (hunk 10 above)
Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/hooks/californiaIdUniqueCheck/useCaliforniaIdUniqueCheck.ts:9]: so it would

File: libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (tglaz) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx]: 🙏

**Review by tglaz — APPROVED** [review]

**Review by shiozakiyuko — COMMENTED** [review]
Hi @mikefeldberg 

I didn't get to see the visual feedback after I submitted a CA ID. I tried again by rescanning this QR code to refresh and start over again but I get the same result. 
The visual feedback: what I meant is after tapping the Submit button, I expected the page guides me to view-only mode of Client Profile page but it didn't happen. 

Although as I tapped the Back button (thinking not saving the ID info), the info I entered was there. So I know it's saved but somehow, it just doesn't behave the way it should be. 

It's happening on both, Android and iOS.

Please check. thanks!

File: libs/expo/betterangels/src/lib/hooks/californiaIdUniqueCheck/ClientCaliforniaUniqueCheck.graphql [hunk 11]
```diff
@@ -0,0 +1,7 @@
+query ClientCaliforniaUniqueCheck($filters: ClientProfileFilter) {
```
Comment (shiozakiyuko) [inline libs/expo/betterangels/src/lib/hooks/californiaIdUniqueCheck/ClientCaliforniaUniqueCheck.graphql:1]: Hi @mikefeldberg

I didn't get to see the visual feedback after I submitted a CA ID. I tried again by rescanning this QR code to refresh and start over again but I get the same result.
The visual feedback: what I meant is after tapping the Submit button, I expected the page guides me to view-only mode of Client Profile page but it didn't happen.

Although as I tapped the Back button (thinking not saving the ID info), the info I entered was there. So I know it's saved but somehow, it just doesn't behave the way it should be.

It's happening on both, Android and iOS.

Please check. thanks!

Comment (shiozakiyuko) [reply libs/expo/betterangels/src/lib/hooks/californiaIdUniqueCheck/ClientCaliforniaUniqueCheck.graphql:1]: Thanks, @mikefeldberg 

I confirmed the submit action takes me to the view-only mode page now. 

One more thing I noticed though.

This is only happening on iOS (Android was fine). 

When I submitted the invalid CA ID, I see the validation message at the beginning. But if I ignored it and tapped on the Submit button anyway, the message disappears. But the error snicker bar (alert) shows up (understandable). My questions is: shouldn't the validation message be persistent till the user enters the valid format or remove the entry?

Thanks!

Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/hooks/californiaIdUniqueCheck/ClientCaliforniaUniqueCheck.graphql:1]: @shiozakiyuko I see what happened here. fixed!

File: libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx [hunk 12]
```diff
@@ -192,27 +192,46 @@ export default function AddEditClient({ id }: { id?: string }) {
...
-        } else {
-          throw new Error(
-            `Failed to update a client profile: ${operationResult.messages[0].message}`
-          );
         }
-      }
 
+        if (
+          resultMessage ===
+          'California ID must be 1 letter followed by 7 numbers'
```
[the remark this answers is not in the document]
Comment (mikefeldberg) [reply libs/expo/betterangels/src/lib/screens/AddEditClient/index.tsx:211]: I moved the uniqueness check into the child component. Tom noted that the client notification will pop up in a modal, which can be triggered from anywhere, so there's no point in keeping this logic higher

I also added the uniqueness check to the field rules. This fixes the behavior that yuko noted [above](https://github.com/BetterAngelsLA/monorepo/pull/705/files#r1841373779)

**Review by Davit-BetterAngels — APPROVED** [review]

**Review by shiozakiyuko — APPROVED** [review]
Looks good to me!
